File: src/main/java/minic/ui/info/MiniCGuideDocument.java
package minic.ui;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Map;

final class MiniCGuideDocument {
    private static final Path DEFAULT_GUIDE = Path.of("docs", "GUIDE.md");
    private static final String DEFAULT_VERSION = "1.0.0";

    private MiniCGuideDocument() {}

    static Path defaultGuidePath() {
        return DEFAULT_GUIDE;
    }

    static String loadDefault() {
        return load(DEFAULT_GUIDE);
    }

    static String load(Path guidePath) {
        String markdown = readGuide(guidePath);
        for (Map.Entry<String, String> entry : runtimeVariables().entrySet()) {
            markdown = markdown.replace("{{" + entry.getKey() + "}}", entry.getValue());
        }
        return markdown;
    }

    private static String readGuide(Path guidePath) {
        try {
            return Files.readString(guidePath, StandardCharsets.UTF_8);
        } catch (IOException exception) {
            return "# MiniC 使用指南\n\nGUIDE.md 未找到: `" + guidePath + "`\n";
        }
    }

    private static Map<String, String> runtimeVariables() {
        LinkedHashMap<String, String> vars = new LinkedHashMap<>();
        vars.put("app.version", appVersion());
        vars.put("java.version", property("java.version"));
        vars.put("java.vendor", property("java.vendor"));
        vars.put("system.os.name", property("os.name"));
        vars.put("system.os.version", property("os.version"));
        vars.put("system.os.arch", property("os.arch"));
        vars.put("system.processors", Integer.toString(Runtime.getRuntime().availableProcessors()));
        return vars;
    }

    private static String appVersion() {
        String configured = System.getProperty("minic.version");
        if (configured != null && !configured.isBlank()) {
            return configured;
        }
        Package pkg = MiniCGuideDocument.class.getPackage();
        String implementation = pkg == null ? null : pkg.getImplementationVersion();
        return implementation == null || implementation.isBlank() ? DEFAULT_VERSION : implementation;
    }

    private static String property(String key) {
        String value = System.getProperty(key);
        return value == null || value.isBlank() ? "unknown" : value;
    }
}

File: uiweb/src/info/MiniCGuideDocument.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import guideMarkdown from "../../../docs/GUIDE.md?raw";

export const miniCGuideDocumentMirror = {
  "javaPath": "src/main/java/minic/uilocal/info/MiniCGuideDocument.java",
  "webPath": "uiweb/src/info/MiniCGuideDocument.ts",
  "packageName": "minic.uilocal",
  "exportName": "MiniCGuideDocument",
  "kind": "class",
  "imports": [
    "java.io.IOException",
    "java.nio.charset.StandardCharsets",
    "java.nio.file.Files",
    "java.nio.file.Path",
    "java.util.LinkedHashMap",
    "java.util.Map"
  ],
  "fields": [
    {
      "name": "DEFAULT_GUIDE",
      "signature": "private static final Path DEFAULT_GUIDE="
    },
    {
      "name": "DEFAULT_VERSION",
      "signature": "private static final String DEFAULT_VERSION="
    }
  ],
  "methods": [
    {
      "name": "appVersion",
      "signature": "appVersion()"
    },
    {
      "name": "property",
      "signature": "property(String key)"
    },
    {
      "name": "readGuide",
      "signature": "readGuide(Path guidePath)"
    },
    {
      "name": "runtimeVariables",
      "signature": "runtimeVariables()"
    }
  ]
} as const satisfies JavaMirrorFile;

export class MiniCGuideDocument {
  static readonly mirror = miniCGuideDocumentMirror;
  static readonly DEFAULT_GUIDE = "docs/GUIDE.md";
  static readonly DEFAULT_VERSION = "1.0.0";

  readonly mirror = miniCGuideDocumentMirror;

  static defaultGuidePath(): string {
    return MiniCGuideDocument.DEFAULT_GUIDE;
  }

  static loadDefault(): string {
    return MiniCGuideDocument.load(MiniCGuideDocument.DEFAULT_GUIDE);
  }

  static load(guidePath = MiniCGuideDocument.DEFAULT_GUIDE): string {
    let markdown = MiniCGuideDocument.readGuide(guidePath);
    for (const [key, value] of Object.entries(MiniCGuideDocument.runtimeVariables())) {
      markdown = markdown.replaceAll(`{{${key}}}`, value);
    }
    return markdown;
  }

  static readGuide(guidePath: string): string {
    if (guidePath === MiniCGuideDocument.DEFAULT_GUIDE) {
      return guideMarkdown;
    }
    return `# MiniC 使用指南\n\nGUIDE.md 未找到: \`${guidePath}\`\n`;
  }

  static runtimeVariables(): Record<string, string> {
    return {
      "app.version": MiniCGuideDocument.appVersion(),
      "java.version": "browser",
      "java.vendor": "browser",
      "system.os.name": MiniCGuideDocument.property("platform"),
      "system.os.version": "browser",
      "system.os.arch": MiniCGuideDocument.property("userAgent"),
      "system.processors": String(navigator.hardwareConcurrency || 1),
    };
  }

  static appVersion(): string {
    return MiniCGuideDocument.DEFAULT_VERSION;
  }

  static property(key: string): string {
    const value = key === "platform" ? navigator.platform : key === "userAgent" ? navigator.userAgent : "";
    return value.trim().length === 0 ? "unknown" : value;
  }

  summary(): string {
    return `MiniCGuideDocument: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }
}

export default MiniCGuideDocument;
